File: maze.java
import java.io.*;
import java.net.*;
import java.util.*;

class Maze {

	public static void main(String args[]) {
		Maze m = new Maze();
		m.main();
	}

	void main() {
		read();
		// dfs(startRow, startCol);
		bfs();
	}

	void bfs() {
		Item first = new Item(startRow, startCol, -1);
		queue.put(first);

		while (!queue.isEmpty()) {
			Item current = queue.get();
			if (current.row == goalRow && current.col == goalCol) {
				for (int i = 0; i < height; i++) {
					for (int j = 0; j < width; j++) {
						if (maze[i][j] == 'O') {
							maze[i][j]  = ' ';
						}
					}
				}
				while (current.parent != -1) {
					maze[current.row][current.col] = 'O';
					current = queue.get(current.parent);
				}
				print();
			} else {
				if (maze[current.row+1][current.col  ] == ' ') {
					maze[current.row+1][current.col  ]  = 'O';
					Item next = new Item(current.row+1, current.col  ,
						queue.front-1);
					queue.put(next);
				}
				if (maze[current.row-1][current.col  ] == ' ') {
					maze[current.row-1][current.col  ]  = 'O';
					Item next = new Item(current.row-1, current.col  ,
						queue.front-1);
					queue.put(next);
				}
				if (maze[current.row  ][current.col+1] == ' ') {
					maze[current.row  ][current.col+1]  = 'O';
					Item next = new Item(current.row  , current.col+1,
						queue.front-1);
					queue.put(next);
				}
				if (maze[current.row  ][current.col-1] == ' ') {
					maze[current.row  ][current.col-1]  = 'O';
					Item next = new Item(current.row  , current.col-1,
						queue.front-1);
					queue.put(next);
				}
			}
		}
	}

	Queue<Item> queue = new Queue<>();

	void dfs(int r, int c) {
		maze[r][c] = 'O';
		if (r == goalRow && c == goalCol) {
			print();
		} else {
			if (maze[r+1][c  ] == ' ') {
				dfs (r+1, c  );
				maze[r+1][c  ]  = ' ';
			}
			if (maze[r-1][c  ] == ' ') {
				dfs (r-1, c  );
				maze[r-1][c  ]  = ' ';
			}
			if (maze[r  ][c+1] == ' ') {
				dfs (r  , c+1);
				maze[r  ][c+1]  = ' ';
			}
			if (maze[r  ][c-1] == ' ') {
				dfs (r  , c-1);
				maze[r  ][c-1]  = ' ';
			}
		}
	}

	void read() {
		maze = new char[MAX][MAX];
		height = 0;
		Scanner scanner = new Scanner(System.in);
		while (scanner.hasNextLine()) {
			String s = scanner.nextLine();
			width = s.length();
			for (int j = 0; j < width; j++) {
				maze[height][j] = s.charAt(j);
			}
			height++;
		}

		for (int i = 0; i < height; i++) {
			for (int j = 0; j < width; j++) {
				if (maze[i][j] == 'R') {
					startRow = i;
					startCol = j;
					maze[i][j] = ' ';
				}
				if (maze[i][j] == 'X') {
					goalRow = i;
					goalCol = j;
					maze[i][j] = ' ';
				}
			}
		}
	}

	void print() {
		for (int i = 0; i < height; i++) {
			for (int j = 0; j < width; j++) {
				System.out.print(maze[i][j]);
			}
			System.out.println();
		}
	}

	int MAX = 256;
	char [][] maze;
	int width;
	int height;
	int startRow = -1, startCol = -1, goalRow = -1, goalCol = -1;
}

class Item {
	int row;
	int col;
	int parent;
	Item(int r, int c, int p) {
		row = r;
		col = c;
		parent = p;
	}
}

class Queue<T> {
	ArrayList<T> queue = new ArrayList<>();
	int front = 0;
	int rear  = 0;
	void put(T t) {
		queue.add(t);
		rear++;
	}

	T get() {
		front++;
		return queue.get(front-1);
	}

	T get(int index) {
		return queue.get(index);
	}

	boolean isEmpty() {
		return front == rear;
	}
}

File: maze.cs
using System;
using System.Collections;
using System.Collections.Generic;

class Maze {

	static void Main() {
		Maze maze = new Maze();
		maze.Read();
		maze.BfsMain();
	}

	void BfsMain() {
		int startRow = -1, startCol = -1;
		int goalRow  = -1, goalCol  = -1;
		for (int i = 0; i < height; i++) {
			for (int j = 0; j < width; j++) {
				if (maze[i,j] == 'R') {
					startRow = i;
					startCol = j;
					maze[i,j] = ' ';
				}
				if (maze[i,j] == 'X') {
					goalRow = i;
					goalCol = j;
					maze[i,j] = ' ';
				}
			}
		}

		Bfs(startRow, startCol, goalRow, goalCol);
	}

	void Bfs(int row, int col, int goalRow, int goalCol) {
		PutQueue(row, col, -1);
		while (!IsQueueEmpty()) {
			QueueItem current = GetQueue();
			Console.WriteLine(current.ToString());

			if (current.row == goalRow && current.col == goalCol) {
				for (int i = 0; i < height; i++) {
					for (int j = 0; j < width; j++) {
						if (maze[i,j] == 'O') {
							maze[i,j] =  ' ';
						}
					}
				}

				while (true) {
					maze[current.row,current.col] = 'O';
					current = queue[current.parent];
					if (current.parent == -1)
						break;
				}

				Print();
				break;
			} else {
				if (maze[current.row+1,current.col  ] == ' ') {
					maze[current.row+1,current.col  ] =  'O';
					PutQueue(current.row+1, current.col  , front-1);
				}
				if (maze[current.row-1,current.col  ] == ' ') {
					maze[current.row-1,current.col  ] =  'O';
					PutQueue(current.row-1, current.col  , front-1);
				}
				if (maze[current.row  ,current.col+1] == ' ') {
					maze[current.row  ,current.col+1] =  'O';
					PutQueue(current.row  , current.col+1, front-1);
				}
				if (maze[current.row  ,current.col-1] == ' ') {
					maze[current.row  ,current.col-1] =  'O';
					PutQueue(current.row  , current.col-1, front-1);
				}
			}
		}
	}

	int front = 0;
	int rear  = 0;
	QueueItem [] queue = new QueueItem[MAX*MAX];
	void PutQueue(int row, int col, int parent) {
		queue[rear] = new QueueItem();
		queue[rear].row = row;
		queue[rear].col = col;
		queue[rear].parent = parent;
		rear++;
	}

	QueueItem GetQueue() {
		front++;
		return queue[front-1];
	}

	bool IsQueueEmpty() {
		return front == rear;
	}

	void DfsMain() {
		int startRow = -1, startCol = -1;
		int goalRow  = -1, goalCol  = -1;
		for (int i = 0; i < height; i++) {
			for (int j = 0; j < width; j++) {
				if (maze[i,j] == 'R') {
					startRow = i;
					startCol = j;
					maze[i,j] = ' ';
				}
				if (maze[i,j] == 'X') {
					goalRow = i;
					goalCol = j;
					maze[i,j] = ' ';
				}
			}
		}

		Dfs(startRow, startCol, goalRow, goalCol);
	}

	void Dfs(int row, int col, int goalRow, int goalCol) {
		maze[row,col] = 'O';
		if (row == goalRow && col == goalCol) {
			Print();
		} else {
			if (maze[row+1,col  ] == ' ') {
				Dfs (row+1,col  , goalRow, goalCol);
				maze[row+1,col  ] =  ' ';
			}
			if (maze[row-1,col  ] == ' ') {
				Dfs (row-1,col  , goalRow, goalCol);
				maze[row-1,col  ] =  ' ';
			}
			if (maze[row  ,col+1] == ' ') {
				Dfs (row  ,col+1, goalRow, goalCol);
				maze[row  ,col+1] =  ' ';
			}
			if (maze[row  ,col-1] == ' ') {
				Dfs (row  ,col-1, goalRow, goalCol);
				maze[row  ,col-1] =  ' ';
			}
		}
	}

	void Read() {
		width = 0;
		height = 0;
		while (true) {
			String s = Console.ReadLine();
			if (s == null) break;
			width = s.Length;
			for (int j = 0; j < width; j++) {
				maze[height, j] = s[j];
			}
			height++;
		}
	}

	void Print() {
		for (int i = 0; i < height; i++) {
			for (int j = 0; j < width; j++) {
				Console.Write(maze[i,j]);
			}
			Console.WriteLine();
		}
	}

	static int MAX = 100;
	char [,] maze = new char[MAX,MAX];
	int width;
	int height;
}

class QueueItem {
	public int row;
	public int col;
	public int parent;
	override public String ToString() {
		return "row=" + row + ", col=" + col + ", parent=" + parent;
	}
}
